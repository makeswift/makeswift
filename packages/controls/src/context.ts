import { z } from 'zod'

import { type Data, type Element, type ElementData } from './common/types'

export type TranslationDto = Record<string, Data>
export type MergeTranslatableDataContext = {
  translatedData: TranslationDto
  mergeTranslatedData: (node: Element) => Element
}

export const removeResourceTagSchema = z.object({
  __type: z.literal('ReplacementContextAction'),
  action: z.literal('remove'),
})

type RemoveResourceTag = z.infer<typeof removeResourceTagSchema>

export function createRemoveTag(): RemoveResourceTag {
  return {
    __type: 'ReplacementContextAction',
    action: 'remove',
  } as const
}

function isRemoveResourceTag(
  value: string | RemoveResourceTag | null | undefined,
): value is RemoveResourceTag {
  return removeResourceTagSchema.safeParse(value).success
}

type ResourceMapping = Map<string, string | RemoveResourceTag>

export type ReplacementContext = {
  elementHtmlIds: Set<string>
  elementKeys: Map<string, string>
  swatchIds: ResourceMapping
  fileIds: ResourceMapping
  typographyIds: ResourceMapping
  tableIds: ResourceMapping
  tableColumnIds: ResourceMapping
  pageIds: ResourceMapping
  globalElementIds: ResourceMapping
  globalElementData: Map<string, ElementData>
}

type SerializableResourceMapping = Record<string, string | RemoveResourceTag>

export type SerializableReplacementContext = {
  elementHtmlIds?: string[]
  elementKeys?: Record<string, string>
  swatchIds?: SerializableResourceMapping
  fileIds?: SerializableResourceMapping
  typographyIds?: SerializableResourceMapping
  tableIds?: SerializableResourceMapping
  tableColumnIds?: SerializableResourceMapping
  pageIds?: SerializableResourceMapping
  globalElementIds?: SerializableResourceMapping
  globalElementData?: Record<string, ElementData>
}

export function createReplacementContext(
  context: SerializableReplacementContext = {},
): ReplacementContext {
  const toMap = <V>(record?: Record<string, V>) =>
    new Map(Object.entries(record ?? {}))

  return {
    elementHtmlIds: new Set(context.elementHtmlIds),
    elementKeys: toMap(context.elementKeys),
    swatchIds: toMap(context.swatchIds),
    fileIds: toMap(context.fileIds),
    typographyIds: toMap(context.typographyIds),
    tableIds: toMap(context.tableIds),
    tableColumnIds: toMap(context.tableColumnIds),
    pageIds: toMap(context.pageIds),
    globalElementIds: toMap(context.globalElementIds),
    globalElementData: toMap(context.globalElementData),
  }
}

export type CopyContext = {
  replacementContext: ReplacementContext
  copyElement: (node: Element) => Element
}

export type MergeContext = {
  mergeElement(a: Element, b: Element): Element
}

function shouldRemoveResource(id: string, map: ResourceMapping): boolean {
  return isRemoveResourceTag(map.get(id))
}

function getReplacementResourceId(
  id: string,
  map: Map<string, string | RemoveResourceTag>,
): string | null {
  const replacement = map.get(id)
  return typeof replacement === 'string' ? replacement : null
}

export function shouldRemoveSwatch(
  swatchId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(swatchId, ctx.replacementContext.swatchIds)
}

export function getReplacementSwatchId(
  swatchId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(swatchId, ctx.replacementContext.swatchIds)
}

export function shouldRemoveFile(fileId: string, ctx: CopyContext): boolean {
  return shouldRemoveResource(fileId, ctx.replacementContext.fileIds)
}

export function getReplacementFileId(
  fileId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(fileId, ctx.replacementContext.fileIds)
}

export function shouldRemoveTypography(
  typographyId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(
    typographyId,
    ctx.replacementContext.typographyIds,
  )
}

export function getReplacementTypographyId(
  typographyId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(
    typographyId,
    ctx.replacementContext.typographyIds,
  )
}

export function shouldRemoveTable(tableId: string, ctx: CopyContext): boolean {
  return shouldRemoveResource(tableId, ctx.replacementContext.tableIds)
}

export function getReplacementTableId(
  tableId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(tableId, ctx.replacementContext.tableIds)
}

export function shouldRemoveTableColumn(
  tableColumnId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(
    tableColumnId,
    ctx.replacementContext.tableColumnIds,
  )
}

export function getReplacementTableColumnId(
  tableColumnId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(
    tableColumnId,
    ctx.replacementContext.tableColumnIds,
  )
}

export function shouldRemovePage(pageId: string, ctx: CopyContext): boolean {
  return shouldRemoveResource(pageId, ctx.replacementContext.pageIds)
}

export function getReplacementPageId(
  pageId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(pageId, ctx.replacementContext.pageIds)
}

export function shouldRemoveGlobalElement(
  globalElementId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(
    globalElementId,
    ctx.replacementContext.globalElementIds,
  )
}

export function getReplacementGlobalElementId(
  globalElementId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(
    globalElementId,
    ctx.replacementContext.globalElementIds,
  )
}
