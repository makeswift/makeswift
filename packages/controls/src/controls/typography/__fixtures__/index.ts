export * from './typography-without-id'
