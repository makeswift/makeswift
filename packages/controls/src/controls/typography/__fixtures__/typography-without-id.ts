export const typographyWithoutId = {
  "style": [
      {
          "deviceId": "desktop",
          "value": {
              "fontFamily": "Nanum Gothic Coding",
              "fontSize": {
                  "unit": "px",
                  "value": 18
              },
              "fontWeight": 400,
              "lineHeight": 1.5
          }
      }
  ]
}
