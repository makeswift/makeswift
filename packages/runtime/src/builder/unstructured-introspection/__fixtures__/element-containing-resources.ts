import { Element } from '@makeswift/controls'

/*
  Below is a fairly minimal example of an Element containing one of each:
    - Swatch
    - File
    - Typography
    - Global Element
    - Page Link (called 'Page' in APIResourceType)
    - Table
*/
export const elementContainingResources: Element = {
  key: '2c5e9ffc-dbcf-486a-ac62-8ac9efa539f0',
  props: {
    children: {
      '@@makeswift/type': 'prop-controllers::grid::v1',
      value: {
        columns: [
          {
            deviceId: 'desktop',
            value: {
              spans: [[12], [12], [12], [12], [12], [12]],
              count: 12,
            },
          },
        ],
        elements: [
          {
            key: 'cb66f6cb-0656-4424-81d7-0d03b27c45de',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[12]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '0f56d678-341a-4b05-bc9c-3a9f5bd7251e',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: swatch only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          color: {
                                            alpha: 1,
                                            swatchId:
                                              'U3dhdGNoOmQ1ZWU3MmY1LThmMTQtNDQzZC05YmQ1LTJiMDA1MTIyYTA5NQ==',
                                          },
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: 'c632090b-2269-40c9-8924-0ecc5effa651',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: 'cf3e1cd1-e1bb-4420-8644-34d780b5465b',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[6, 6]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: 'c42ddde2-b9ff-4787-909f-d23866ca9130',
                      props: {
                        file: {
                          '@@makeswift/type': 'prop-controllers::image::v2',
                          value: {
                            id: 'RmlsZTpjNzc2MjI3ZC1iNmZjLTQ3MGYtYTY1OC00NWQ4NzA2NDVkMDM=',
                            type: 'makeswift-file',
                            version: 1,
                          },
                        },
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginLeft: {
                                  unit: 'px',
                                  value: 270,
                                },
                                marginRight: 'auto',
                              },
                            },
                          ],
                        },
                        width: {
                          '@@makeswift/type': 'prop-controllers::width::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                unit: '%',
                                value: 15,
                              },
                            },
                          ],
                        },
                      },
                      type: './components/Image/index.js',
                    },
                    {
                      key: 'ce4edb81-8832-4266-a152-4e506e1e754d',
                      props: {
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginBottom: {
                                  unit: 'px',
                                  value: 20,
                                },
                                marginLeft: 'auto',
                                marginRight: 'auto',
                                marginTop: null,
                              },
                            },
                          ],
                        },
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: File only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '061f43b1-72aa-46b6-b429-c4a55eac09c5',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: 'ceebe1f0-31c1-4314-b387-df463d076751',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[12]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '87ab03ef-3a10-49d9-b596-719834f894dd',
                      props: {
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginBottom: {
                                  unit: 'px',
                                  value: 20,
                                },
                                marginLeft: 'auto',
                                marginRight: 'auto',
                                marginTop: null,
                              },
                            },
                          ],
                        },
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: typography only',
                                  typography: {
                                    id: 'VHlwb2dyYXBoeTo1ODVkMGE1ZS0zYmUyLTQyNzEtOTczYS1mNWNjN2UzMzk4ODg=',
                                    style: [],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '9c7f5fa6-7f15-4e60-91fd-778827ab3449',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: '6f2a2437-f690-4644-84d6-14073f1404d4',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[6, 6]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '42a6048c-ea36-41a4-9783-9c28f58fa3e6',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: Global Element only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '09c570fa-7d82-4886-8c33-ad418202d562',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                    {
                      key: '229fc932-2fdf-46eb-8ff6-ad31d3660d0a',
                      type: 'reference',
                      value: 'R2xvYmFsRWxlbWVudDphOTRjYjM3ZS02OWM2LTQ1OGQtODIxNS1mZTA2NGQ3OWZmNjI=',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: '3c83ae93-7cc9-47b1-8599-4efbd5bbd331',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[6, 6]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: 'f1ee5bf3-ebec-4398-b4e1-73b13c8050dd',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: Page Link only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: 'f3b4e0db-8141-4500-8b0d-19978cfcd664',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                    {
                      key: '0ca675bf-a128-4a22-9cda-b4f07ad7b52e',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: '',
                                },
                                {
                                  children: [
                                    {
                                      text: 'Sample Link to Page A',
                                      typography: {
                                        style: [
                                          {
                                            deviceId: 'desktop',
                                            value: {
                                              fontSize: {
                                                unit: 'px',
                                                value: 18,
                                              },
                                              fontWeight: 400,
                                              lineHeight: 1.5,
                                            },
                                          },
                                          {
                                            deviceId: 'mobile',
                                            value: {
                                              fontSize: {
                                                unit: 'px',
                                                value: 16,
                                              },
                                            },
                                          },
                                        ],
                                      },
                                    },
                                  ],
                                  link: {
                                    payload: {
                                      openInNewTab: false,
                                      pageId:
                                        'UGFnZToxZWU1YmNjNi1mMGI1LTQxMGUtOWIzMy0wZTJjZTNkYTIwMzE=',
                                    },
                                    type: 'OPEN_PAGE',
                                  },
                                  type: 'link',
                                },
                                {
                                  text: '',
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '878ec264-dbb5-4b14-8dc3-651133ba9a8e',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: 'fac5fb69-c058-49a8-8b07-0917b2beddc8',
            props: {
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        spans: [[6, 6]],
                        count: 12,
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '9274acff-aa56-4f04-b4f2-04e8106dc3a0',
                      props: {
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginBottom: {
                                  unit: 'px',
                                  value: 20,
                                },
                                marginLeft: {
                                  unit: 'px',
                                  value: 0,
                                },
                                marginRight: 'auto',
                                marginTop: null,
                              },
                            },
                          ],
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                        text: {
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: table',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '78307876-10bc-4240-a01d-071cd395a586',
                        },
                      },
                      type: './components/Text/index.js',
                    },
                    {
                      key: 'f5fbb12a-948b-414d-8267-b571713cbf60',
                      type: './components/Form/index.js',
                      props: {
                        gap: [
                          {
                            deviceId: 'desktop',
                            value: {
                              value: 10,
                              unit: 'px',
                            },
                          },
                        ],
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              value: 550,
                              unit: 'px',
                            },
                          },
                        ],
                        tableId: {
                          '@@makeswift/type': 'prop-controllers::table::v1',
                          value: 'VGFibGU6MDI2OTUwMDQtOThkMi00NDU1LThjNDAtYmFiMDU5YTY3NDQy',
                        },
                        fields: {
                          '@@makeswift/type': 'prop-controllers::table-form-fields::v1',
                          value: {
                            fields: [],
                            grid: [
                              {
                                deviceId: 'desktop',
                                value: {
                                  count: 12,
                                  spans: [[12]],
                                },
                              },
                            ],
                          },
                        },
                      },
                    },
                  ],
                },
              },
            },
            type: './components/Box/index.js',
          },
        ],
      },
    },
  },
  type: './components/Root/index.js',
}
