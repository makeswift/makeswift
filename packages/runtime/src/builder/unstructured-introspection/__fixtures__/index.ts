export * from './element-containing-resources'
