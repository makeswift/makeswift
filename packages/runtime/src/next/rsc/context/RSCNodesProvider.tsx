'use client'

import { createContext, ReactNode, useContext } from 'react'

type RSCNodes = Record<string, JSX.Element>

const RSCNodesContext = createContext<RSCNodes>({})

export const RSCNodesProvider = ({ children, value }: { children: ReactNode; value: RSCNodes }) => {
  return <RSCNodesContext.Provider value={value}>{children}</RSCNodesContext.Provider>
}

export const useRSCNodes = () => {
  return useContext(RSCNodesContext)
}
