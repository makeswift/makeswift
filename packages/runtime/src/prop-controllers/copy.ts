import { type CopyContext } from '@makeswift/controls'

import {
  copyLinkPropControllerData,
  copyShadowsPropControllerData,
  copyBorderPropControllerData,
  copyElementIDPropControllerData,
  copyGridPropControllerData,
  copyImagePropControllerData,
  copyImagesPropControllerData,
  copyBackgroundsPropControllerData,
  copyResponsiveColorPropControllerData,
  copyTablePropControllerData,
  copyNavigationLinksPropControllerData,
  Types as PropControllerTypes,
} from '@makeswift/prop-controllers'

import { isLegacyDescriptor, type Descriptor } from './descriptors'

// @note: note typing value, because would then have to type narrow `Data` per case
export function copy(descriptor: Descriptor, value: any, context: CopyContext) {
  if (!isLegacyDescriptor(descriptor)) {
    return descriptor.copyData(value, context)
  }

  switch (descriptor.type) {
    case PropControllerTypes.Backgrounds:
      return copyBackgroundsPropControllerData(descriptor, value, context)

    case PropControllerTypes.Grid:
      return copyGridPropControllerData(value, context)

    case PropControllerTypes.NavigationLinks:
      return copyNavigationLinksPropControllerData(value, context)

    case PropControllerTypes.Link:
      return copyLinkPropControllerData(value, context)

    case PropControllerTypes.Shadows:
      return copyShadowsPropControllerData(value, context)

    case PropControllerTypes.Image:
      return copyImagePropControllerData(value, context)

    case PropControllerTypes.Images:
      return copyImagesPropControllerData(value, context)

    case PropControllerTypes.ResponsiveColor:
      return copyResponsiveColorPropControllerData(value, context)

    case PropControllerTypes.TableFormFields:
      return copyTablePropControllerData(value, context)

    case PropControllerTypes.Table:
      return copyTablePropControllerData(value, context)

    case PropControllerTypes.Border:
      return copyBorderPropControllerData(value, context)

    case PropControllerTypes.ElementID:
      return copyElementIDPropControllerData(value, context)

    // FIXME: DELETED_PROP_CONTROLLER_TYPES.RichText ??

    default:
      return value
  }
}
