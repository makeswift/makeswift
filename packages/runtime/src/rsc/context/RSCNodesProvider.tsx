'use client'

import { createContext, ReactNode, useContext, useEffect } from 'react'
import { setRSCElementKeys } from '../../state/actions'
import { useDispatch } from '../../runtimes/react/hooks/use-dispatch'

type RSCNodes = Record<string, JSX.Element>

const RSCNodesContext = createContext<RSCNodes>({})

export const RSCNodesProvider = ({ children, value }: { children: ReactNode; value: RSCNodes }) => {
  const dispatch = useDispatch()

  useEffect(() => {
    const keys = Object.keys(value)
    dispatch(setRSCElementKeys(keys))
  }, [value, dispatch])

  return <RSCNodesContext.Provider value={value}>{children}</RSCNodesContext.Provider>
}

export const useRSCNodes = () => {
  return useContext(RSCNodesContext)
}
