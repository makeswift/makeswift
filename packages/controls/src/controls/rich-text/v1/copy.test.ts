import { createReplacementContext } from '../../../context'

import * as Fixtures from '../__fixtures__'

import { copyRichTextData } from './copy'

describe('GIVEN copying RichText', () => {
  test('replaces the swatch, typography and page ids', () => {
    const expected = JSON.parse(
      JSON.stringify(Fixtures.introspection)
        .replace(Fixtures.SWATCH_ID, 'replaced-swatch-id')
        .replace(Fixtures.TYPOGRAPHY_ID, 'replaced-typography-id')
        .replace(Fixtures.PAGE_ID, 'replaced-page-id'),
    )

    const result = copyRichTextData(Fixtures.introspection, {
      replacementContext: createReplacementContext({
        swatchIds: { [Fixtures.SWATCH_ID]: 'replaced-swatch-id' },
        typographyIds: { [Fixtures.TYPOGRAPHY_ID]: 'replaced-typography-id' },
        pageIds: { [Fixtures.PAGE_ID]: 'replaced-page-id' },
      }),
      copyElement: (node) => node,
    })

    expect(result).toMatchObject(expected)
  })
})
