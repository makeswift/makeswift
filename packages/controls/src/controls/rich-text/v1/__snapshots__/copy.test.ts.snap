// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN copying RichText removes any swatch, typography and page ids marked for removal 1`] = `
{
  "annotations": undefined,
  "data": undefined,
  "document": {
    "data": undefined,
    "nodes": [
      {
        "data": {},
        "nodes": [
          {
            "data": {
              "payload": {
                "openInNewTab": false,
                "pageId": undefined,
              },
              "type": "OPEN_PAGE",
            },
            "nodes": [
              {
                "marks": [
                  {
                    "data": {
                      "value": undefined,
                    },
                    "object": "mark",
                    "type": "typography",
                  },
                ],
                "object": "text",
                "text": "Credibly implement global synergy, then collaboratively implement goal-oriented sprints. Rapidiously initiate standards compliant clouds. ",
              },
            ],
            "object": "inline",
            "type": "link",
          },
        ],
        "object": "block",
        "type": "blockquote",
      },
    ],
    "object": "document",
  },
  "object": "value",
  "selection": {
    "anchor": {
      "object": "point",
      "offset": 0,
      "path": [
        0,
        0,
      ],
    },
    "focus": {
      "object": "point",
      "offset": 0,
      "path": [
        1,
        2,
      ],
    },
    "isFocused": false,
    "marks": undefined,
    "object": "selection",
  },
}
`;
